symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -211,43 +172,55 @@ instance
     type WitnessField (ArithmeticCircuit a p i) = WitnessF a (WitVar p i)
     witnessF (behead -> (c, o)) = o <&> \case
       ConstVar cv -> fromConstant cv
-      SysVar (InVar iv) -> at $ SysVar (InVar iv)
-      SysVar (NewVar nv) -> acWitness c ! nv
+      LinVar k (InVar iv) b -> at $ LinVar k (InVar iv) b
+      LinVar k (NewVar nv) b -> fromConstant k * acWitness c ! nv + fromConstant b
     fromCircuitF (behead -> (c, o)) f = uncurry (set #acOutput) (runState (f o) c)
 
 ----------------------------- MonadCircuit instance ----------------------------
 
 instance Finite a => Witness (Var a i) (WitnessF a (WitVar p i)) where
-  at (ConstVar cV) = fromConstant cV
-  at (SysVar sV)   = WitnessF (\x -> x (WSysVar sV))
+  at (ConstVar cV)   = fromConstant cV
+  at (LinVar k sV b) = fromConstant k * pure (WSysVar sV) + fromConstant b
 
 instance
   ( Arithmetic a, Binary a, Binary (Rep p), Binary (Rep i), Ord (Rep i)
   , o ~ U1) => MonadCircuit (Var a i) a (WitnessF a (WitVar p i)) (State (ArithmeticCircuit a p i o)) where
 
-    unconstrained wf = case runWitnessF wf $ const Nothing of
-      Just cV -> return (ConstVar cV)
-      Nothing -> do
-        let v = toVar @a wf
-        -- TODO: forbid reassignment of variables
-        zoom #acWitness $ modify (insert v wf)
-        return $ SysVar (NewVar v)
+    unconstrained wf =
+        case runWitnessF wf $ \case
+          WSysVar sV -> LinUVar one sV zero
+          _          -> More
+        of
+          LinUVar k x b -> return (LinVar k x b)
+          _ -> do
+            let v = witToVar @a wf
+            -- TODO: forbid reassignment of variables
+            zoom #acWitness $ modify (insert v wf)
+            return $ toVar (NewVar v)
 
     constraint p =
       let evalConstVar = \case
-            SysVar sysV -> var sysV
+            LinVar k sysV b -> fromConstant k * var sysV + fromConstant b
             ConstVar cV -> fromConstant cV
           evalMaybe = \case
-            SysVar _ -> Nothing
             ConstVar cV -> Just cV
+            _ -> Nothing
       in case p evalMaybe of
-        Just c -> if c ==zero
+        Just c -> if c == zero
                     then return ()
                     else error "The constraint is non-zero"
-        Nothing -> zoom #acSystem . modify $ insert (toVar @_ @p (p at)) (p evalConstVar)
-
-    rangeConstraint (SysVar v) upperBound =
-      zoom #acRange . modify $ insertWith S.union upperBound (S.singleton v)
+        Nothing -> zoom #acSystem . modify $ insert (witToVar @_ @p (p at)) (p evalConstVar)
+
+    rangeConstraint (LinVar k x b) upperBound = if k == one && b == zero
```

Comment:
I'd extract the operation of "prepare `SysVar` which is suitable for range constraint" in a separate function.
Also the case `k == -1 && b == upperBound` is also ok here.